test/zipkin/util.go

```diff
@@ -93,17 +95,33 @@ func SetupZipkinTracing(kubeClientset *kubernetes.Clientset, logf logging.Format
 		logf("Successfully setup SpoofingClient for Zipkin Tracing")
 		ZipkinTracingEnabled = true
 	})
+	return ZipkinTracingEnabled
 }
 
 // CleanupZipkinTracingSetup cleans up the Zipkin tracing setup on the machine. This involves killing the process performing port-forward.
+// This should be called exactly once in TestMain. Likely in the form:
+//
+// func TestMain(m *testing.M) {
+//     os.Exit(wrappedMain(m))
+// }
+//
+// func wrappedMain(m *testing.M) {
+//    // Any setup required for the tests.
+//    defer zipkin.CleanupZipkinTracingSetup(logger)
+//    return m.Run()
+// }
 func CleanupZipkinTracingSetup(logf logging.FormatLogger) {
 	teardownOnce.Do(func() {
+		// Because CleanupZipkinTracingSetup only runs once, make sure that now that it has been
+		// run, SetupZipkinTracing will no longer setup any port forwarding.
+		setupOnce.Do(func() {})
```

Comment:
Not sure if this does what you want?

Reply:
My reading of sync.Once is that it will allow Do() to be run at most once. So there are three cases:
1. Something has already called Do(). Then this is a no-op. It will return quickly.
2. Something is currently calling Do(). Then this will block until that call has finished. Thus ensuring we clean things up appropriately.
3. Nothing has ever called Do(). Then this will ensure nothing can do so in the future.

I'm pretty sure this does what I want, doesn't it?

Reply:
But your `Do` is a no-op...

Reply:
There are two sync.Onces, `teardownOnce` and `setupOnce`. If `teardownOnce` is being called, we need to make sure that `setupOnce` is not called afterwards. That is what this line is doing, by doing this no-op, anyone calling `zipkin.SetupZipkinTracing` in the future will _not_ create a proxy-forwarding process.